src/ppom-storage.ts

```diff
@@ -0,0 +1,208 @@
+ */
+export class PPOMStorage {
+  readonly #storageBackend: StorageBackend;
+
+  readonly #readMetadata: () => PPOMFileMetadata;
+
+  readonly #writeMetadata: (metadata: PPOMFileMetadata) => void;
+
+  /**
+   * Creates a PPOMStorage instance.
+   *
+   * @param options - The options passed to the function.
+   * @param options.storageBackend - The storage backend to use for the local storage.
+   * @param options.readMetadata - A function to read the metadata from the local storage.
+   * @param options.writeMetadata - A function to write the metadata to the local storage.
+   */
+  constructor({
+    storageBackend,
+    readMetadata,
+    writeMetadata,
+  }: {
+    storageBackend: StorageBackend;
+    readMetadata: () => PPOMFileMetadata;
+    writeMetadata: (metadata: PPOMFileMetadata) => void;
+  }) {
+    this.#storageBackend = storageBackend;
+    this.#readMetadata = readMetadata;
+    this.#writeMetadata = writeMetadata;
+  }
+
+  /**
+   * Sync the metadata with the version info from the cdn.
+   * 1. Remove the files that are not readable (e.g. corrupted or deleted).
+   * 2. Remove the files that are not in the cdn anymore.
+   * 3. Remove the files that are not up to date in the cdn.
+   * 4. Remove the files that are not in the local storage from the metadata.
+   * 5. Delete the files that are not in the metadata from the local storage.
+   *
+   * @param versionInfo - Version information of metadata files.
+   */
+  async syncMetadata(versionInfo: FileInfo[]): Promise<PPOMFileMetadata> {
+    const metadata = this.#readMetadata();
+    const syncedMetadata = [];
+
+    for (const fileMetadata of metadata) {
+      // check if the file is readable (e.g. corrupted or deleted)
+      try {
+        await this.readFile(fileMetadata.name, fileMetadata.chainId);
+      } catch (exp: any) {
+        continue;
+      }
+
+      // check if the file exits and up to date in the storage
+      if (
+        !versionInfo.find(
+          (file) =>
+            file.name === fileMetadata.name &&
+            file.chainId === fileMetadata.chainId &&
+            file.version === fileMetadata.version &&
+            file.checksum === fileMetadata.checksum,
+        )
+      ) {
+        continue;
+      }
+
+      syncedMetadata.push(fileMetadata);
+    }
+
+    const filesInDB = await this.#storageBackend.dir();
+    for (const { name, chainId } of filesInDB) {
+      if (
+        !syncedMetadata.find(
+          (file) => file.name === name && file.chainId === chainId,
+        )
+      ) {
+        await this.#storageBackend.delete({ name, chainId });
+      }
+    }
+
+    this.#writeMetadata(syncedMetadata);
+    return syncedMetadata;
+  }
+
+  /**
+   * Read the file from the local storage.
+   * 1. Check if the file exists in the local storage.
+   * 2. Check if the file exists in the metadata.
+   * 3. Check if the checksum is valid.
+   *
+   * @param name - Name assigned to storage.
```

Comment:
would be nice to add the param type where possible
```suggestion
   * @param {string} name - Name assigned to storage.
```